src/Stratis.Bitcoin/Connection/PeerLifetime.cs

```diff
@@ -0,0 +1,145 @@
+    /// </summary>
+    public interface IBanStore
+    {
+        /// <summary>
+        /// Set a peer endpoint to a banned state.
+        /// </summary>
+        /// <param name="endpoint">The endpoint to set that it was banned.</param>
+        /// <param name="banUntil">The date when the ban will expire.</param>
+        void BanPeer(IPEndPoint endpoint, DateTime banUntil);
+
+        /// <summary>
+        /// Get information about a banned peer.
+        /// </summary>
+        /// <param name="endpoint">The endpoint to query.</param>
+        /// <returns>The <see cref="DateTime"/> of the end of the ban or null if the peer was not found.</returns>
+        DateTime? TryGetBannedPeer(IPEndPoint endpoint);
+    }
+
+    /// <summary>
+    /// A temporary memory store for banned peers.
+    /// </summary>
+    public class MemoryBanStore : IBanStore
+    {
+        private readonly ConcurrentDictionary<IPAddress, DateTime> banned = new ConcurrentDictionary<IPAddress, DateTime>();
+
+        /// <inheritdoc />
+        public void BanPeer(IPEndPoint endpoint, DateTime banUntil)
+        {
+            this.banned.AddOrUpdate(endpoint.Address, banUntil, (point, time) => banUntil);
+        }
+
+        /// <inheritdoc />
+        public DateTime? TryGetBannedPeer(IPEndPoint endpoint)
+        {
+            this.banned.TryGetValue(endpoint.Address, out DateTime banUntil);
+
+            return banUntil;
+        }
+    }
+
+    /// <summary>
+    /// A class that manages the lifetime of a peer.
+    /// </summary>
+    public class PeerLifetime : IPeerLifetime
+    {
+        /// <summary>A connection manager of peers.</summary>
+        private readonly IConnectionManager connectionManager;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>A store of banned peers.</summary>
+        private readonly IBanStore banStore;
+
+        /// <summary>Functionality of date and time.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>Functionality of date and time.</summary>
+        private readonly ConnectionManagerSettings connectionManagerSettings;
+
+        public PeerLifetime(IConnectionManager connectionManager, ILoggerFactory loggerFactory, IDateTimeProvider dateTimeProvider, NodeSettings nodeSettings)//, IBanStore banStore)
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.connectionManager = connectionManager;
+            this.banStore =  new MemoryBanStore();// banStore;
+            this.dateTimeProvider = dateTimeProvider;
+            this.connectionManagerSettings = nodeSettings.ConnectionManager;
+        }
+
+        /// <inheritdoc />
+        public void BanPeer(IPEndPoint endpoint, string reason = null)
+        {
+            Guard.NotNull(endpoint, nameof(endpoint));
+
+            this.logger.LogTrace("()");
```

Comment:
`this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(endpoint), endpoint, nameof(reason), reason);`